cpp/include/cudf/detail/iterator.cuh

```diff
@@ -80,6 +81,104 @@ struct null_replaced_value_accessor
+ * if it is valid, or `pair(null_replacement, false)` if it is null.
+ *
+ * @throws `cudf::logic_error` if the column is not nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+template <typename Element>
+struct null_replaced_pair_accessor: public null_replaced_value_accessor<Element>
+{
+  /** -------------------------------------------------------------------------*
+   * @brief constructor
+   * @param[in] _col column device view of cudf column
+   * @param[in] null_replacement The value to return for null elements
+   * -------------------------------------------------------------------------**/
+  null_replaced_pair_accessor(column_device_view const& _col, Element null_val)
+      : null_replaced_value_accessor<Element>{_col, null_val}
+  { }
+
+  CUDA_DEVICE_CALLABLE
+  thrust::pair<Element, bool> operator()(cudf::size_type i) const {
+    return this->col.is_valid_nocheck(i)
+               ? thrust::make_pair(this->col.template element<Element>(i), true)
+               : thrust::make_pair(this->null_replacement, false);
+  }
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief pair accessor of column with/without null bitmask
+ * A unary functor returns pair with scalar value at `id` and boolean validity
+ * `operator() (cudf::size_type id)` computes `element`  and
+ * returns a `pair(element, validity)`
+ *
+ * the return value for element `i` will return `pair(column[i], validity)`
+ * `validity` is `true` if `has_nulls=false`.
+ * `validity` is validity of the element at `i` if `has_nulls=true` and the
+ * column is nullable.
+ *
+ * @throws `cudf::logic_error` if `has_nulls==true` and the column is not
+ * nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @tparam has_nulls boolean indicating to treat the column is nullable
+ * -------------------------------------------------------------------------**/
+template <typename Element, bool has_nulls=false> 
+struct pair_accessor
+{
+  column_device_view const col;      ///< column view of column in device
+
+  /** -------------------------------------------------------------------------*
+   * @brief constructor
+   * @param[in] _col column device view of cudf column
+   * -------------------------------------------------------------------------**/
+  pair_accessor(column_device_view const& _col)
+      : col{_col}
+  {
+    CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == col.type(),
+                 "the data type mismatch");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  thrust::pair<Element, bool> operator()(cudf::size_type i) const {
+    return {col.element<Element>(i), true};
+  }
+};
+// @brief specialization of `pair_accessor` for nullable column
```

Comment:
Since the implementation is otherwise the same, why not branch on `has_nulls` instead of having a specialized template for it?

Reply:
Yes. That's better. merged them both.